drivers: ethernet: vsc8541: various fixes

robhancocksed opened the pull request:
- Added missing MDIO enable/disable calls around accessing PHY registers, to make sure the MDIO bus is active first
- Cleaned up internal functions to use uint16_t for PHY register values
- Fixed inverted reset GPIO line
- Added SW reset timeout
- Implemented cfg_link function required by some MAC drivers

robhancocksed committed:
drivers: ethernet: phy: vsc8541: add MDIO enable/disable

The driver was not enabling the MDIO bus before trying to access
registers. Added enabling and disabling the bus around PHY register
accesses.

Signed-off-by: Robert Hancock <robert.hancock@calian.com>

robhancocksed committed:
drivers: ethernet: phy: vsc8541: Use 16-bit values for MDIO access

The internal register read/write functions used uint32_t for the values
even though the registers are only 16 bits wide, resulting in a bunch of
casting. Change the internal functions to use uint16_t and wrap them for
the external read/write API which uses uint32_t.

Signed-off-by: Robert Hancock <robert.hancock@calian.com>

robhancocksed committed:
drivers: ethernet: vsc8541: Fixed inverted reset GPIO

For GPIOs driving active-low signals, such as the VSC8541's reset pin,
they are supposed to be declared as active low in the device tree, and
set to 1 to assert and 0 to clear. Change the driver as such so that it
does not leave the PHY stuck in reset when so configured.

Also changed all in-tree board DTS files for this PHY to properly
declare the reset GPIO as active low.

Signed-off-by: Robert Hancock <robert.hancock@calian.com>

robhancocksed committed:
drivers: ethernet: phy: vsc8541: Add timeout on SW reset

The driver previously could enter an infinite loop if the PHY software
reset failed to complete, which could happen due to hardware reset
issues or MDIO bus problems. Add a timeout of 1000 iterations so we
report an error in this scenario rather than causing a lockup.

Signed-off-by: Robert Hancock <robert.hancock@calian.com>

robhancocksed commented:
It looks like the arch.interrupt.gen_isr_table_local.riscv test is failing on the ganymed_bob and ganymed_sk boards where I modified the .dts files. However, those build errors also seem to happen on the main branch, so they seem unrelated to these changes?

robhancocksed committed:
doc: migration guide: ethernet: Mention vsc8541 reset changes

Mention a change to the reset-gpios handling in the vsc8541 PHY driver.

Signed-off-by: Robert Hancock <robert.hancock@calian.com>

robhancocksed commented on boards/sensry/ganymed_bob/ganymed_bob_sy120_gbm.dts:
```diff
@@ -30,7 +30,7 @@
 		status = "okay";
 
-		reset-gpios = <&gpio0 11 GPIO_ACTIVE_HIGH>;
+		reset-gpios = <&gpio0 11 GPIO_ACTIVE_LOW>;
```
Added

robhancocksed commented on drivers/ethernet/phy/phy_microchip_vsc8541.c:
```diff
@@ -216,10 +216,15 @@ static int phy_mc_vsc8541_reset(const struct device *dev)
+		if (count++ > 1000) {
+			LOG_ERR("phy reset timed out");
+			return -ETIMEDOUT;
+		}
```
Done

maass-hamburg committed:
drivers: ethernet: phy: microchip_vsc8541: use mutex

use mutex to protect page register

phy_mc_vsc8541_get_link got removed from
phy_mc_vsc8541_link_cb_set so, that
phy_mc_vsc8541_link_monitor (own thread)
is the only one to change data->state

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
drivers: ethernet: phy: microchip_vsc8541: improve driver

- implement configure link
- support half duplex
- use defines from mii.h
- fix check ret vals

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg commented on drivers/ethernet/phy/phy_microchip_vsc8541.c:
```diff
@@ -400,13 +405,80 @@ static int phy_mc_vsc8541_get_link(const struct device *dev, struct phy_link_sta
+ * @param adv_speeds speeds to be advertised
  */
-static int phy_mc_vsc8541_cfg_link(const struct device *dev, enum phy_link_speed speeds)
+static int phy_mc_vsc8541_cfg_link(const struct device *dev, enum phy_link_speed adv_speeds)
```
Feel free to cherry-pick https://github.com/zephyrproject-rtos/zephyr/pull/91572/commits/b58b241474f02ad53c03c8055a18b84ed7966759 from https://github.com/zephyrproject-rtos/zephyr/pull/91572
and putting it before your changes

I fixed there some other things I noticed

robhancocksed committed:
drivers: ethernet: phy: vsc8541: fixed build warnings

Fixed some build warnings in the driver from previous changes by
removing an unused variable and hooking up the cfg_link function. Also
removes some implicit boolean conversions.

Signed-off-by: Robert Hancock <robert.hancock@calian.com>

robhancocksed committed:
drivers: ethernet: phy: vsc8541: allow disabling autonegotiation

Add support for disabling autonegotiation to the cfg_link callback, as
with the phy_mii driver.

Signed-off-by: Robert Hancock <robert.hancock@calian.com>

robhancocksed commented on drivers/ethernet/phy/phy_microchip_vsc8541.c:
```diff
@@ -271,72 +227,83 @@ static int phy_mc_vsc8541_reset(const struct device *dev)
+{
+	int ret;
+
+	if (flags & PHY_FLAG_AUTO_NEGOTIATION_DISABLED) {
```
Fixed these

maass-hamburg commented on drivers/ethernet/phy/phy_microchip_vsc8541.c:
```diff
@@ -204,62 +185,37 @@ static int phy_mc_vsc8541_reset(const struct device *dev)
+			LOG_ERR("phy reset timed out");
+			return -ETIMEDOUT;
+		}
+	} while (reg & MII_BMCR_RESET);
```
No implicit boolean conversions

robhancocksed replied:
Fixed this one as well.

robhancocksed commented:
> @robhancocksed please rebase to current main, hopefully the unrelated CI fails had been fixed, if not please open an issue

As you noted, the test failures on ganymed_bob are also occurring on the main branch, so rebasing will not help at this point. I assume this PR is flagging those failures since it touched the .dts files for those boards. So either this can be merged despite the test failures, or https://github.com/zephyrproject-rtos/zephyr/issues/92194 will need to be fixed.

maass-hamburg commented:
> > @robhancocksed please rebase to current main, hopefully the unrelated CI fails had been fixed, if not please open an issue
> 
> As you noted, the test failures on ganymed_bob are also occurring on the main branch, so rebasing will not help at this point. I assume this PR is flagging those failures since it touched the .dts files for those boards. So either this can be merged despite the test failures, or #92194 will need to be fixed.

correct, probably needs a fix like https://github.com/zephyrproject-rtos/zephyr/pull/91391 but for your effected boards

dkalowsk commented:
@robhancocksed RC1 has been tagged, which means all new PRs need to be bug fixes with an attached GitHub issue.  Given that this PR is all bug fixes, can you create an issue and attach it to this PR so we can merge for RC2?

robhancocksed commented:
Rebased to main to pick up the change in https://github.com/zephyrproject-rtos/zephyr/pull/92336 which should hopefully fix the CI build error.

maass-hamburg commented:
@robhancocksed this needs a rebase to current main

robhancocksed commented:
Rebased to main and moved the migration guide entry to 4.3 now that 4.2 has been released.
